### PR TITLE
Reinstate `:member_of_collection_id` argument option to Actor Stack 

### PR DESCRIPTION
**Pull request by no-reply**
Refs #2545, #2533.

This merge includes some minor refactors of `#create`; including some changes to
the actor's private methods.

`#extract_collection_ids` and `#assign_nested_attributes_for_collection` now
accept only the `env` as an argument. `#assign_nested_attributes_for_collection`
ensures that the `env` is left in the correct state. This reduces the impact of
mutable state across methods.

Assigning the singleton `#collection_id` is still a separate mutation of
`env`. This may be necessary without a much more extensive refactor.

`#extract_collection_id` is updated to handle the old syntax, which
ensures that `#create` populates the `:collection_id` environment attribute in
the same way as it does for the new syntax. Likewise, `#valid_membership?` is
updated to accept the collection ids it should check; existing calls are changed
to pass the ids directly.

Changes proposed in this pull request:
* Merge commit '089bff9e8f8100db6119f5' into collections-sprint
* Support new collections features for old collections_id actor syntax

@samvera/hyrax-code-reviewers

### CODE REVIEW

File: app/actors/hyrax/actors/collections_membership_actor.rb [hunk 1]
```diff
@@ -75,19 +121,18 @@ def has_destroy_flag?(hash)
...
         # Given an array of collection_attributes when it is size:
         # * 0 do not set `env.attributes[:collection_id]`
         # * 1 set `env.attributes[:collection_id]` to the one and only one collection
         # * 2+ do not set `env.attributes[:collection_id]`
         #
         # NOTE: Only called from create.  All collections are being added as parents of a work.  None are being removed.
-        def extract_collection_id(env, attributes_collection)
+        def extract_collection_id(env)
+          attributes_collection =
+            env.attributes.fetch(:member_of_collections_attributes) { nil }
```
Comment (elrayle) [inline app/actors/hyrax/actors/collections_membership_actor.rb]: This needs to be able to extract the id from the new :member_of_collections_attributes OR from the old style :member_of_collection_ids.  

See diffs for method extract_collection_id (line 84) in the PR that switched from the old style :member_of_collection_ids to the new :member_of_collections_attributes...  https://github.com/samvera/hyrax/commit/85591cafa9e9935d18fd3b07cf86fb35cb9a73d4#diff-dcaa34b138208d26b5cc4a005542ad71

File: spec/actors/hyrax/actors/collections_membership_actor_spec.rb [hunk 2]
```diff
@@ -100,6 +100,52 @@
...
+            Hyrax::VERSION.split('.').first.to_i >= 3
+
+          expect(subject.create(env)).to be true
+          expect(curation_concern.member_of_collections)
+            .to contain_exactly(collection, my_collection)
+        end
+      end
+    end
+
     context "updates env" do
```
Comment (elrayle) [inline spec/actors/hyrax/actors/collections_membership_actor_spec.rb]: Need to also update the tests under `context "updated env"` to test that the update to env happens for both :member_of_collections_attributes and :member_of_collection_ids parameters.

Effected tests:  https://github.com/samvera/hyrax/blob/9c5fa41dd2e161ac70421aae573e0dea3c7a8e77/spec/actors/hyrax/actors/collections_membership_actor_spec.rb#L149-L203

Comment (elrayle) [reply spec/actors/hyrax/actors/collections_membership_actor_spec.rb]: You can copy in the tests that use the old style parameter from here...
https://github.com/samvera/hyrax/blob/d67f598e992bcf5897c0731db1551ca8698cb04e/spec/actors/hyrax/actors/collections_membership_actor_spec.rb#L89-L129

**Comment by no-reply** [comment]
@elrayle See the questions in the PR description.

It's not clear to me what the best way to handle the new behavior for `member_of_collection_ids`. In particular, adding the `#valid_membership?` check would be breaking for DCE's current use cases, which rather defeats the purpose.

Some additional context on the motivation for the `#valid_membership?` behavior, and how `collection_id` is intended to be used would help me determine the best implementation for the old syntax.

**Comment by no-reply** [comment]
It might help as well if the `collection_id` attribute were named more explicitly. What behavior would I expect as a caller to the actor stack if I pass such an id directly, or extract it in another actor, either before or after this?

**Comment by elrayle** [comment]
Questions - answered

> Do we need to handle extraction of the #collection_id attribute in the #member_of_collection_ids case?

Yes.  It was handled before the change was made to member_of_collection_ids.  You can see the code that processed it prior to the change at https://github.com/samvera/hyrax/blob/d67f598e992bcf5897c0731db1551ca8698cb04e/app/actors/hyrax/actors/collections_membership_actor.rb#L40-L51

This older code does not include additional changes that are seen in the current code that need to be handled now for both parameters.

> Does multiple membership need to be treated as invalid under #member_of_collection_ids, as well as in the new attributes syntax?

Yes.  See description of multiple membership below next question.

> Should we be concerned that the current head on collection-sprint prevents adding multiple collections at the actor level? What is the reasoning for this restriction? Am I misunderstanding #valid_membership?

There is `multiple_membership` configuration for collection types.  #valid_membership will only deny multiple collections if that configuration is false.  The variable could probably be better named `multiple_membership_denied`.  `Hyrax::MultipleMembershipChecker` will only return a value if multiple collections are being added AND `multiple_membership=false`.
https://github.com/samvera/hyrax/blob/9c5fa41dd2e161ac70421aae573e0dea3c7a8e77/app/services/hyrax/multiple_membership_checker.rb#L33

If you look at that service, you will see how the test is applied.  The multiple membership test only applies to collections of the same type.

> It might help as well if the collection_id attribute were named more explicitly. What behavior would I expect as a caller to the actor stack if I pass such an id directly, or extract it in another actor, either before or after this?

Because this actor deletes the env parameter, the collection that effects permissions has to be added into the env.  To avoid confusion, a different name was given.  The `collection_id` parameter is processed by a later actor in the stack to apply permissions... 
https://github.com/samvera/hyrax/blob/collections-sprint/app/actors/hyrax/actors/apply_permission_template_actor.rb#L26-L31

**Comment by no-reply** [comment]
I think the info you've given is enough for me to finish this work up.

I'm still of the opinion that the current PR supports the status quo on both branches, and can be safely merged. Though perhaps the ticket needs to be left open to track the remaining compatibility work. If you disagree, I can wrap this up first thing in the morning and we can move forward.

----

> Because this actor deletes the env parameter, the collection that effects permissions has to be added into the env. To avoid confusion, a different name was given. The collection_id parameter is processed by a later actor in the stack to apply permissions...

Would it be more accurate to rename it `permission_template_source` or similar? I'm concerned a bit about how tightly coupled these two actors are---I know this kind of thing is endemic in the actor stack, and I'm not too interested in wringing my hands about it, but some explict naming would help a lot, I think.

In either case, I think this is a separate issue. If it seems worth further discussion to you, I can open it after merge.

**Comment by elrayle** [comment]
@no-reply Would you consider a simpler change that would ensure that all functionality is properly applied regardless of which parameter format is used.

You can replace line 15 and 24...
```
attributes_collection = env.attributes.delete(:member_of_collections_attributes)
```
with a method call that extracts the values from env...
```
attributes_collection = process_env_attribute(env)
```
and add method...
```
# Extract env attribute that specifies collection membership processing.  There are two supported formats:
# @example a collections attribute hash - New style that allows for adding and deleting collection membership.
#   'member_of_collections_attributes' => {
#     '0' => { 'id' = '12312412'},
#     '1' => { 'id' = '234234234'},
#     '2' => { 'id' = '99981228', '_destroy' => 'true' }
#   }
#
# @example a collection ids hash - Old style that allows for adding collection membership.  This will be supported through Hyrax 3.0 for backward compatibility.
#   'member_of_collection_ids' => { '12312412', '234234234' }
def process_env_attribute(env)
  # return new env attribute approach if it exists
  return env.attributes.delete(:member_of_collections_attributes) if env.attributes.key? :member_of_collections_attributes

  # convert old style env attribute into new style supporting backward compatibility
  return nil unless env.attributes.key? :member_of_collection_ids
  collection_ids = env.attributes.delete(:member_of_collection_ids)
  return nil if collection_ids.blank?
  attributes = {}
  0.upto(collection_ids.size-1) { |i| attributes[i.to_s] = { 'id' => collection_ids[i] }
  attributes
end
```

**Comment by elrayle** [comment]
I make this suggestion, because on first review, I missed that you did not include all the new functionality that was put in place in the [#assign_collections method](https://github.com/samvera/hyrax/blob/d67f598e992bcf5897c0731db1551ca8698cb04e/app/actors/hyrax/actors/collections_membership_actor.rb#L22-L34) as written before the switch to the new style env parameter.  I am hesitant to see the processing of each style parameter happen completely separate because it opens up the possibility of something getting missed and causes duplicate code.

File: app/actors/hyrax/actors/collections_membership_actor.rb [hunk 3]
```diff
@@ -46,6 +52,46 @@ def assign_nested_attributes_for_collection(env, attributes_collection)
...
+        def emit_deprecation
+          Deprecation.warn(self, ':member_of_collections_attributes and :member_of_collection_ids were both ' \
+                                 ' passed. :member_of_collection_ids is ignored when both are passed and is ' \
+                                 'deprecated for removal in Hyrax 3.0.')
+        end
+
+        ##
+        # @deprecated supports old :member_of_collection_ids arguments
+        def assign_for_collection_ids(env)
+          collection_ids = env.attributes.delete(:member_of_collection_ids)
```
Comment (elrayle) [inline app/actors/hyrax/actors/collections_membership_actor.rb]: The assign_for_collection_ids before the switch to the new style env parameter was...
```
        def assign_collections(env, collection_ids)
          return true unless collection_ids
          multiple_memberships = Hyrax::MultipleMembershipChecker.new(item: env.curation_concern).check(collection_ids: collection_ids)
          if multiple_memberships
            env.curation_concern.errors.add(:collections, multiple_memberships)
            return false
          end
          # grab/save collections this user has no edit access to
          other_collections = collections_without_edit_access(env)
          env.curation_concern.member_of_collections = ::Collection.find(collection_ids)
          env.curation_concern.member_of_collections.concat other_collections
end
```

I don't have a problem with the addition of the exception, but this method needs to include the check for multiple members.

**Comment by elrayle** [comment]
Seems like there are minimally two pieces of work for this PR to be merged:
* add in the multi_membership check for the old style env parameter and 
* copy in most of the tests that were in place before the env parameter change from https://github.com/samvera/hyrax/blob/d67f598e992bcf5897c0731db1551ca8698cb04e/spec/actors/hyrax/actors/collections_membership_actor_spec.rb#L36-L129

**Comment by no-reply** [comment]
@elrayle I've force pushed with the requested changes.

I think you will probably want some more refactoring here (I certainly do), but I'm inclined (again) to push forward with this and take on the remainder after the collections merge. I've also run into various potential test suite issues, where stubs are being set incorrectly or ineffectually, which has me questioning how effective the unit tests on this class are. I would like to address any refactors and test suite fixes in follow-up work.

**Comment by no-reply** [comment]
> Need to open a new issue to revisit difference in processing for the two style env parameters. We should be able to reformat the old style into the new style and use a single approach for processing. If this isn't sufficient, then we need to understand why.

I honestly think I prefer the branched logic to increased mutations of `env`. But I do agree that there are more refactors to be had here. My goals would be to:

**Comment by no-reply** [comment]
The biggest issue I see in the current code is that both `#extract_collection_id` and `#assign_nested_attributes_for_collection` need to read the same information from `env`. I have a notion to  add a flag parameter to `#assign_nested_attributes_for_collection` like `extract_permission_source_id` (is this an informative name for what `extract_collection_id` does?). This way we can extract the data once, then pass it along to `#extract_collection_id`. Something like:

**Pull request merged**